File: 2017/11/hex-ed.py
# -*- coding: utf-8 -*-
import sys


DIRECTIONS = {
    'n'  : ( 0,  1, -1),  # noqa
    's'  : ( 0, -1,  1),  # noqa
    'ne' : ( 1,  0, -1),  # noqa
    'sw' : (-1,  0,  1),  # noqa
    'nw' : (-1,  1,  0),  # noqa
    'se' : ( 1, -1,  0),  # noqa
}


def read_directions(filename):
    with open(filename) as f:
        data = f.read().strip()

    return data.split(',')


def move(point, offset):
    return tuple(sum(a) for a in zip(point, offset))


def follow(directions, start=(0, 0, 0)):
    coord = start
    for direction in directions:
        coord = move(coord, DIRECTIONS[direction])
    return coord


def get_distance(start, end):
    return max(sum((a, -b)) for a, b in zip(start, end))


def main(filename):
    center = 0, 0, 0
    directions = read_directions(filename)
    location = follow(directions, start=center)
    distance = get_distance(location, center)
    print(distance)


if __name__ == '__main__':
    main(sys.argv[1])

File: 2017/11/hex-ed.go
package main

import "fmt"
import "io/ioutil"
import "os"
import "strings"


type Point [3]int


var DIRECTIONS = map[string]Point{
    "n" : [3]int{ 0,  1, -1},
    "s" : [3]int{ 0, -1,  1},
    "ne": [3]int{ 1,  0, -1},
    "sw": [3]int{-1,  0,  1},
    "nw": [3]int{-1,  1,  0},
    "se": [3]int{ 1, -1,  0},
}


var CENTER = [3]int{0, 0, 0}


func readDirections(filename string) []string {
    data, err := ioutil.ReadFile(filename)
    if err != nil {
        panic(err)
    }

    var contents = strings.TrimSpace(string(data))
    return strings.Split(contents, ",")
}


func follow(directions []string, start Point) Point {
    var coord = start
    for _, direction := range directions {
        coord = move(coord, DIRECTIONS[direction])
    }
    return coord
}


func move(point Point, offset Point) (result Point) {
    result[0] = point[0] + offset[0]
    result[1] = point[1] + offset[1]
    result[2] = point[2] + offset[2]
    return
}


func getDistance(end Point, start Point) (distance int) {
    for i := 0; i < 3; i += 1 {
        var difference = end[i] - start[i]
        if difference > distance {
            distance = difference
        }
    }
    return
}


func main() {
   var filename = os.Args[1]
   var directions = readDirections(filename)
   var location = follow(directions, CENTER)
   var distance = getDistance(location, CENTER)
   fmt.Println(distance)
}